### DISCUSSION TITLE
Create namespace team1/app1 where app1 is a child of team1

### DISCUSSION THREAD

**Discussion by aymousa** [hvac/hvac#1037] [General]
Hello everyone,  

I am following the step by step and hopefully this does help. 

1. working on a python end to end script that will execute and use import hvac during py functions
2. I am using HashiCorp vault enterprise development for testing the requirements use case
3. I have mast namespace under vault root named devops
4. the goal is to automate multi-tenant as follow
5. create new child namespace using `client2.sys.create_namespace(path="app1)`
6. the code in line line 5 hardcode namespace which I am looking for a way to change it to a string parameter running python command cli but it is not working getting python errors
7. below the full code block 
```python
import hvac

client = hvac.Client(url='https://vault-URL/ui/vault/secrets?namespace=**devops**')

# Create namespace daa/new-child-namespacename where the new namespace will be used for a specific tenant is a child of devops/
client2 = hvac.Client(url="https://vault-URL", namespace="devops")
client2.sys.create_namespace(path="NEW namespace")
```

8. I tried using `:type namespace_path: str` it did not work
9. I rewrote the code block as following and still not working as it is erroring out see latest code block 
```python
import os
import hvac
import requests

def hashivault_argspec():
    argument_spec = dict(
        namespace=dict(required=False, default=os.environ.get('VAULT_NAMESPACE', None), type='str')
    )
    return argument_spec

def hashivault_client(params):
    url = params.get('url')
    check_verify = params.get('verify')
    namespace = params.get('namespace', None)
    if check_verify == '' or check_verify:
        if url:
            verify = url
        else:
            verify = check_verify
        client = hvac.Client(url=url, verify=verify, namespace=namespace)
        return client
```

**Comment by amousa1968** [comment]
Does this seems to be the correct way doing it: 

import os
import hvac
import requests

def hashivault_argspec():
    argument_spec = dict(
        namespace=dict(required=False, default=os.environ.get('VAULT_NAMESPACE', None), type='str')
    )
    return argument_spec

def hashivault_client(params):
    url = params.get('url')
    check_verify = params.get('verify')
    namespace = params.get('namespace', None)
    if check_verify == '' or check_verify:
        if url:
            verify = url
    else:
        verify = check_verify
    client = hvac.Client(url=url, verify=verify, namespace=namespace)
    return client

**Comment by amousa1968** [comment]
> Hello @wandermonk
Can I get some help from you
> 
> I need help with hvac library client2.sys.create_namespace(path="**User Input from the command line**"
> 
> I using hvac library in my python code following this example:
> 
> 1. I would like to create child vault namespace under a default namespace but I don't want to hard code the new child namespace in the script using the following using **client2.sys.create_namespace(path="app1")**  as you can see path="app1" is hard coded and I would like to use one of the following options have the library create each time code runs create new dynamic name or prompt the user to input the name from the command line when executing python script.  I am looking for a way to not hard code path="app1". Than you

**Reply by amousa1968** [reply]
I am trying to use hvac.client use param string and not hard code the child namespace from the above script. I 'd like to use python script_name.py Enter a child namespace.    using below example remove app1 and use command line input 
client2.sys.create_namespace(path="app1")

**Comment by briantist** [comment]
@aymousa / @amousa1968 : it's a little hard to follow this discussion, do you think you could move all of your questions into the original post so it's all together? Also please use code blocks for any code you're posting, it's quite difficult to read, especially without the correct indentation. Thank you!

**Reply by aymousa** [reply]
Hi @briantist 

I am following the step by step and hopefully this does help. 

1. working on a python end to end script that will execute and use import hvac during py functions
2. I am using HashiCorp vault enterprise development for testing the requirements use case
3. I have mast namespace under vault root named devops
4. the goal is to automate multi-tenant as follow
5. create new child namespace using `client2.sys.create_namespace(path="app1)`
6. the code in line line 5 hardcode namespace which I am looking for a way to change it to a string parameter running python command cli but it is not working getting python errors
7. below the full code block 
`import hvac

client = hvac.Client(url='https://vault-URL/ui/vault/secrets?namespace=**devops**')

# Create namespace daa/new-child-namespacename where the new namespace will be used for a specific tenant is a child of devops/
client2 = hvac.Client(url="https://vault-URL", namespace="devops")
client2.sys.create_namespace(path="NEW namespace")
`
8. I tried using `:type namespace_path: str` it did not work
9. I rewrote the code block as following and still not working as it is erroring out see latest code block `import os
import hvac
import requests

def hashivault_argspec():
argument_spec = dict(
namespace=dict(required=False, default=os.environ.get('VAULT_NAMESPACE', None), type='str')
)
return argument_spec

def hashivault_client(params):
url = params.get('url')
check_verify = params.get('verify')
namespace = params.get('namespace', None)
if check_verify == '' or check_verify:
if url:
verify = url
else:
verify = check_verify
client = hvac.Client(url=url, verify=verify, namespace=namespace)
return client`

Does this work?

**Reply by briantist** [reply]
Unfortunately the formatting in this post is still very messed up. I've tried to fix it, and I put it back into the original post.

Please edit your original post and refrain from posting additional comments and replies with edits.

Let's use the replies for discussion, thanks!

**Comment by amousa1968** [comment]
Hi Brian,  sure I will reformat the ask. Thanks,Ayman On Aug 31, 2023, at 5:52 AM, Brian Scholer ***@***.***> wrote:﻿
@aymousa / @amousa1968 : it's a little hard to follow this discussion, do you think you could move all of your questions into the original post so it's all together? Also please use code blocks for any code you're posting, it's quite difficult to read, especially without the correct indentation. Thank you!

**Comment by briantist** [comment]
So with the edited post, I'm still a bit confused about what you're trying to do. 

In the second code block, the `hashivault_argspec` function doesn't seem to be used at all in the example code so I'm not sure whether it's relevant.

In the first code block, you create `client` but then use a variable called `client2` and I can't see where or how it's defined.

> it is erroring out see latest code block

What is the error you're getting? How is the code called?
